<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html lang="en">
<head>
<title>Source code</title>
<link rel="stylesheet" type="text/css" href="../../../../../stylesheet.css" title="Style">
</head>
<body>
<div class="sourceContainer">
<pre><span class="sourceLineNo">001</span>/*<a name="line.1"></a>
<span class="sourceLineNo">002</span><a name="line.2"></a>
<span class="sourceLineNo">003</span>    ngs-tools  Next generation sequencing (NGS/HTS) command line tools.<a name="line.3"></a>
<span class="sourceLineNo">004</span>    Copyright (c) 2014 National Marrow Donor Program (NMDP)<a name="line.4"></a>
<span class="sourceLineNo">005</span><a name="line.5"></a>
<span class="sourceLineNo">006</span>    This library is free software; you can redistribute it and/or modify it<a name="line.6"></a>
<span class="sourceLineNo">007</span>    under the terms of the GNU Lesser General Public License as published<a name="line.7"></a>
<span class="sourceLineNo">008</span>    by the Free Software Foundation; either version 3 of the License, or (at<a name="line.8"></a>
<span class="sourceLineNo">009</span>    your option) any later version.<a name="line.9"></a>
<span class="sourceLineNo">010</span><a name="line.10"></a>
<span class="sourceLineNo">011</span>    This library is distributed in the hope that it will be useful, but WITHOUT<a name="line.11"></a>
<span class="sourceLineNo">012</span>    ANY WARRANTY; with out even the implied warranty of MERCHANTABILITY or<a name="line.12"></a>
<span class="sourceLineNo">013</span>    FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public<a name="line.13"></a>
<span class="sourceLineNo">014</span>    License for more details.<a name="line.14"></a>
<span class="sourceLineNo">015</span><a name="line.15"></a>
<span class="sourceLineNo">016</span>    You should have received a copy of the GNU Lesser General Public License<a name="line.16"></a>
<span class="sourceLineNo">017</span>    along with this library;  if not, write to the Free Software Foundation,<a name="line.17"></a>
<span class="sourceLineNo">018</span>    Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA.<a name="line.18"></a>
<span class="sourceLineNo">019</span><a name="line.19"></a>
<span class="sourceLineNo">020</span>    &gt; http://www.gnu.org/licenses/lgpl.html<a name="line.20"></a>
<span class="sourceLineNo">021</span><a name="line.21"></a>
<span class="sourceLineNo">022</span>*/<a name="line.22"></a>
<span class="sourceLineNo">023</span>package org.nmdp.ngs.tools;<a name="line.23"></a>
<span class="sourceLineNo">024</span><a name="line.24"></a>
<span class="sourceLineNo">025</span>import static com.google.common.base.Preconditions.checkArgument;<a name="line.25"></a>
<span class="sourceLineNo">026</span>import static com.google.common.base.Preconditions.checkNotNull;<a name="line.26"></a>
<span class="sourceLineNo">027</span><a name="line.27"></a>
<span class="sourceLineNo">028</span>import static org.dishevelled.compress.Readers.reader;<a name="line.28"></a>
<span class="sourceLineNo">029</span>import static org.dishevelled.compress.Writers.writer;<a name="line.29"></a>
<span class="sourceLineNo">030</span><a name="line.30"></a>
<span class="sourceLineNo">031</span>import java.io.BufferedReader;<a name="line.31"></a>
<span class="sourceLineNo">032</span>import java.io.File;<a name="line.32"></a>
<span class="sourceLineNo">033</span>import java.io.IOException;<a name="line.33"></a>
<span class="sourceLineNo">034</span>import java.io.PrintWriter;<a name="line.34"></a>
<span class="sourceLineNo">035</span><a name="line.35"></a>
<span class="sourceLineNo">036</span>import java.util.ArrayList;<a name="line.36"></a>
<span class="sourceLineNo">037</span>import java.util.Collections;<a name="line.37"></a>
<span class="sourceLineNo">038</span>import java.util.Comparator;<a name="line.38"></a>
<span class="sourceLineNo">039</span>import java.util.HashMap;<a name="line.39"></a>
<span class="sourceLineNo">040</span>import java.util.List;<a name="line.40"></a>
<span class="sourceLineNo">041</span>import java.util.Map;<a name="line.41"></a>
<span class="sourceLineNo">042</span><a name="line.42"></a>
<span class="sourceLineNo">043</span>import java.util.concurrent.Callable;<a name="line.43"></a>
<span class="sourceLineNo">044</span><a name="line.44"></a>
<span class="sourceLineNo">045</span>import com.google.common.base.Joiner;<a name="line.45"></a>
<span class="sourceLineNo">046</span>import com.google.common.base.Splitter;<a name="line.46"></a>
<span class="sourceLineNo">047</span><a name="line.47"></a>
<span class="sourceLineNo">048</span>import com.google.common.collect.ArrayListMultimap;<a name="line.48"></a>
<span class="sourceLineNo">049</span>import com.google.common.collect.ListMultimap;<a name="line.49"></a>
<span class="sourceLineNo">050</span><a name="line.50"></a>
<span class="sourceLineNo">051</span>import org.dishevelled.commandline.ArgumentList;<a name="line.51"></a>
<span class="sourceLineNo">052</span>import org.dishevelled.commandline.CommandLine;<a name="line.52"></a>
<span class="sourceLineNo">053</span>import org.dishevelled.commandline.CommandLineParseException;<a name="line.53"></a>
<span class="sourceLineNo">054</span>import org.dishevelled.commandline.CommandLineParser;<a name="line.54"></a>
<span class="sourceLineNo">055</span>import org.dishevelled.commandline.Switch;<a name="line.55"></a>
<span class="sourceLineNo">056</span>import org.dishevelled.commandline.Usage;<a name="line.56"></a>
<span class="sourceLineNo">057</span><a name="line.57"></a>
<span class="sourceLineNo">058</span>import org.dishevelled.commandline.argument.DoubleArgument;<a name="line.58"></a>
<span class="sourceLineNo">059</span>import org.dishevelled.commandline.argument.IntegerArgument;<a name="line.59"></a>
<span class="sourceLineNo">060</span>import org.dishevelled.commandline.argument.FileArgument;<a name="line.60"></a>
<span class="sourceLineNo">061</span>import org.dishevelled.commandline.argument.StringArgument;<a name="line.61"></a>
<span class="sourceLineNo">062</span><a name="line.62"></a>
<span class="sourceLineNo">063</span>import org.nmdp.ngs.feature.Allele;<a name="line.63"></a>
<span class="sourceLineNo">064</span>import org.nmdp.ngs.feature.AlleleException;<a name="line.64"></a>
<span class="sourceLineNo">065</span>import org.nmdp.ngs.feature.Locus;<a name="line.65"></a>
<span class="sourceLineNo">066</span><a name="line.66"></a>
<span class="sourceLineNo">067</span>import org.nmdp.ngs.feature.parser.FeatureParser;<a name="line.67"></a>
<span class="sourceLineNo">068</span>import org.nmdp.ngs.feature.parser.ParseException;<a name="line.68"></a>
<span class="sourceLineNo">069</span><a name="line.69"></a>
<span class="sourceLineNo">070</span>import htsjdk.samtools.SAMFileReader;<a name="line.70"></a>
<span class="sourceLineNo">071</span>import htsjdk.samtools.SAMRecord;<a name="line.71"></a>
<span class="sourceLineNo">072</span>import htsjdk.samtools.CigarElement;<a name="line.72"></a>
<span class="sourceLineNo">073</span><a name="line.73"></a>
<span class="sourceLineNo">074</span>import org.biojava.bio.seq.DNATools;<a name="line.74"></a>
<span class="sourceLineNo">075</span><a name="line.75"></a>
<span class="sourceLineNo">076</span>import org.biojava.bio.symbol.Edit;<a name="line.76"></a>
<span class="sourceLineNo">077</span>import org.biojava.bio.symbol.IllegalSymbolException;<a name="line.77"></a>
<span class="sourceLineNo">078</span>import org.biojava.bio.symbol.SymbolList;<a name="line.78"></a>
<span class="sourceLineNo">079</span><a name="line.79"></a>
<span class="sourceLineNo">080</span>/**<a name="line.80"></a>
<span class="sourceLineNo">081</span> * Filter consensus sequences into subregions of research or clinical interest.<a name="line.81"></a>
<span class="sourceLineNo">082</span> */<a name="line.82"></a>
<span class="sourceLineNo">083</span>public final class FilterConsensus implements Callable&lt;Integer&gt; {<a name="line.83"></a>
<span class="sourceLineNo">084</span>    private final File inputBamFile;<a name="line.84"></a>
<span class="sourceLineNo">085</span>    private final File inputGenomicFile;<a name="line.85"></a>
<span class="sourceLineNo">086</span>    private final File outputFile;<a name="line.86"></a>
<span class="sourceLineNo">087</span>    private final String gene;<a name="line.87"></a>
<span class="sourceLineNo">088</span>    private final boolean cdna;<a name="line.88"></a>
<span class="sourceLineNo">089</span>    private final boolean removeGaps;<a name="line.89"></a>
<span class="sourceLineNo">090</span>    private final double minimumBreadth;<a name="line.90"></a>
<span class="sourceLineNo">091</span>    private final int expectedPloidy;<a name="line.91"></a>
<span class="sourceLineNo">092</span>    static final double DEFAULT_MINIMUM_BREADTH = 0.5d;<a name="line.92"></a>
<span class="sourceLineNo">093</span>    static final int DEFAULT_EXPECTED_PLOIDY = 2;<a name="line.93"></a>
<span class="sourceLineNo">094</span>    static final String USAGE = "ngs-filter-consensus -i input.bam -x genome.txt -g HLA-A [args]";<a name="line.94"></a>
<span class="sourceLineNo">095</span><a name="line.95"></a>
<span class="sourceLineNo">096</span><a name="line.96"></a>
<span class="sourceLineNo">097</span>    /**<a name="line.97"></a>
<span class="sourceLineNo">098</span>     * Filter consensus sequences into subregions of research or clinical interest.<a name="line.98"></a>
<span class="sourceLineNo">099</span>     *<a name="line.99"></a>
<span class="sourceLineNo">100</span>     * @param inputBamFile input BAM file, must not be null<a name="line.100"></a>
<span class="sourceLineNo">101</span>     * @param inputGenomicFile input genomic region file, must not be null<a name="line.101"></a>
<span class="sourceLineNo">102</span>     * @param outputFile output FASTA file, if any<a name="line.102"></a>
<span class="sourceLineNo">103</span>     * @param gene gene, must not be null<a name="line.103"></a>
<span class="sourceLineNo">104</span>     * @param cdna cdna, true to output cdna from the same contig (phased consensus sequence)<a name="line.104"></a>
<span class="sourceLineNo">105</span>     *    in FASTA format (for interpretation)<a name="line.105"></a>
<span class="sourceLineNo">106</span>     * @param removeGaps remove alignment gaps in the filtered consensus sequence<a name="line.106"></a>
<span class="sourceLineNo">107</span>     * @param minimumBreadth minimum breadth, must be in range [0.0 - 1.0]<a name="line.107"></a>
<span class="sourceLineNo">108</span>     * @param expectedPloidy expected ploidy, must be &amp;gt; 0<a name="line.108"></a>
<span class="sourceLineNo">109</span>     */<a name="line.109"></a>
<span class="sourceLineNo">110</span>    public FilterConsensus(final File inputBamFile,<a name="line.110"></a>
<span class="sourceLineNo">111</span>                           final File inputGenomicFile,<a name="line.111"></a>
<span class="sourceLineNo">112</span>                           final File outputFile,<a name="line.112"></a>
<span class="sourceLineNo">113</span>                           final String gene,<a name="line.113"></a>
<span class="sourceLineNo">114</span>                           final boolean cdna,<a name="line.114"></a>
<span class="sourceLineNo">115</span>                           final boolean removeGaps,<a name="line.115"></a>
<span class="sourceLineNo">116</span>                           final double minimumBreadth,<a name="line.116"></a>
<span class="sourceLineNo">117</span>                           final int expectedPloidy) {<a name="line.117"></a>
<span class="sourceLineNo">118</span>        checkNotNull(inputBamFile);<a name="line.118"></a>
<span class="sourceLineNo">119</span>        checkNotNull(inputGenomicFile);<a name="line.119"></a>
<span class="sourceLineNo">120</span>        checkNotNull(gene);<a name="line.120"></a>
<span class="sourceLineNo">121</span>        checkArgument(minimumBreadth &gt;= 0.0d &amp;&amp; minimumBreadth &lt;= 1.0d, "minimum breadth must be in range [0.0 - 1.0]");<a name="line.121"></a>
<span class="sourceLineNo">122</span>        checkArgument(expectedPloidy &gt; 0, "expected ploidy must be &gt; 0");<a name="line.122"></a>
<span class="sourceLineNo">123</span>        this.inputBamFile = inputBamFile;<a name="line.123"></a>
<span class="sourceLineNo">124</span>        this.inputGenomicFile = inputGenomicFile;<a name="line.124"></a>
<span class="sourceLineNo">125</span>        this.outputFile = outputFile;<a name="line.125"></a>
<span class="sourceLineNo">126</span>        this.gene = gene;<a name="line.126"></a>
<span class="sourceLineNo">127</span>        this.cdna = cdna;<a name="line.127"></a>
<span class="sourceLineNo">128</span>        this.removeGaps = removeGaps;<a name="line.128"></a>
<span class="sourceLineNo">129</span>        this.minimumBreadth = minimumBreadth;<a name="line.129"></a>
<span class="sourceLineNo">130</span>        this.expectedPloidy = expectedPloidy;<a name="line.130"></a>
<span class="sourceLineNo">131</span>    }<a name="line.131"></a>
<span class="sourceLineNo">132</span><a name="line.132"></a>
<span class="sourceLineNo">133</span><a name="line.133"></a>
<span class="sourceLineNo">134</span>    @Override<a name="line.134"></a>
<span class="sourceLineNo">135</span>    public Integer call() throws Exception {<a name="line.135"></a>
<span class="sourceLineNo">136</span>        PrintWriter writer = null;<a name="line.136"></a>
<span class="sourceLineNo">137</span>        try {<a name="line.137"></a>
<span class="sourceLineNo">138</span>            writer = writer(outputFile);<a name="line.138"></a>
<span class="sourceLineNo">139</span><a name="line.139"></a>
<span class="sourceLineNo">140</span>            // map of region alleles from genomic file keyed by "exon" as integer<a name="line.140"></a>
<span class="sourceLineNo">141</span>            Map&lt;Integer, Allele&gt; exons = readGenomicFile(inputGenomicFile);<a name="line.141"></a>
<span class="sourceLineNo">142</span><a name="line.142"></a>
<span class="sourceLineNo">143</span>            // todo:  refactor this block into separate static methods<a name="line.143"></a>
<span class="sourceLineNo">144</span>            // map of overlapping alignment alleles from BAM file keyed by "exon" as integer<a name="line.144"></a>
<span class="sourceLineNo">145</span>            ListMultimap&lt;Integer, Allele&gt; regions = ArrayListMultimap.create();<a name="line.145"></a>
<span class="sourceLineNo">146</span><a name="line.146"></a>
<span class="sourceLineNo">147</span>            for (SAMRecord record : new SAMFileReader(inputBamFile)) {<a name="line.147"></a>
<span class="sourceLineNo">148</span>                List&lt;Edit&gt; edits = cigarToEditList(record);<a name="line.148"></a>
<span class="sourceLineNo">149</span><a name="line.149"></a>
<span class="sourceLineNo">150</span>                String referenceName = record.getReferenceName();<a name="line.150"></a>
<span class="sourceLineNo">151</span>                int start = record.getAlignmentStart();<a name="line.151"></a>
<span class="sourceLineNo">152</span>                int end = record.getAlignmentEnd();<a name="line.152"></a>
<span class="sourceLineNo">153</span>                Locus alignment = new Locus(referenceName, start, end);<a name="line.153"></a>
<span class="sourceLineNo">154</span>                SymbolList sequence = DNATools.createDNA(record.getReadString());<a name="line.154"></a>
<span class="sourceLineNo">155</span><a name="line.155"></a>
<span class="sourceLineNo">156</span>                for (Edit edit : edits) {<a name="line.156"></a>
<span class="sourceLineNo">157</span>                    sequence.edit(edit);<a name="line.157"></a>
<span class="sourceLineNo">158</span>                }<a name="line.158"></a>
<span class="sourceLineNo">159</span><a name="line.159"></a>
<span class="sourceLineNo">160</span>                String name = record.getReadName();<a name="line.160"></a>
<span class="sourceLineNo">161</span>                Allele contig = Allele.builder()<a name="line.161"></a>
<span class="sourceLineNo">162</span>                    .withContig(referenceName)<a name="line.162"></a>
<span class="sourceLineNo">163</span>                    .withStart(start)<a name="line.163"></a>
<span class="sourceLineNo">164</span>                    .withEnd(end)<a name="line.164"></a>
<span class="sourceLineNo">165</span>                    .withSequence(sequence)<a name="line.165"></a>
<span class="sourceLineNo">166</span>                    .build();<a name="line.166"></a>
<span class="sourceLineNo">167</span><a name="line.167"></a>
<span class="sourceLineNo">168</span>                for (Map.Entry&lt;Integer, Allele&gt; entry : exons.entrySet()) {<a name="line.168"></a>
<span class="sourceLineNo">169</span>                    int index = entry.getKey();<a name="line.169"></a>
<span class="sourceLineNo">170</span>                    Allele exon = entry.getValue();<a name="line.170"></a>
<span class="sourceLineNo">171</span><a name="line.171"></a>
<span class="sourceLineNo">172</span>                    String chr = exon.getContig();<a name="line.172"></a>
<span class="sourceLineNo">173</span>                    int min = exon.getMin();<a name="line.173"></a>
<span class="sourceLineNo">174</span>                    int max = exon.getMax();<a name="line.174"></a>
<span class="sourceLineNo">175</span>                    String range = chr + ":" + min + "-" + max;<a name="line.175"></a>
<span class="sourceLineNo">176</span><a name="line.176"></a>
<span class="sourceLineNo">177</span>                    if (alignment.overlaps(exon)) {<a name="line.177"></a>
<span class="sourceLineNo">178</span>                        Allele xover = exon.doubleCrossover(contig);<a name="line.178"></a>
<span class="sourceLineNo">179</span><a name="line.179"></a>
<span class="sourceLineNo">180</span>                        Allele clipped = xover.leftHardClip("-").rightHardClip("-");<a name="line.180"></a>
<span class="sourceLineNo">181</span>                        clipped.setName("&gt;" + name + "|gene=" + gene + "|exon=" + index + "|location=" + range + "|" + (max - min));<a name="line.181"></a>
<span class="sourceLineNo">182</span>                        regions.put(index, clipped);<a name="line.182"></a>
<span class="sourceLineNo">183</span><a name="line.183"></a>
<span class="sourceLineNo">184</span>                        int offset = 0;<a name="line.184"></a>
<span class="sourceLineNo">185</span>                        for (Edit edit : edits) {<a name="line.185"></a>
<span class="sourceLineNo">186</span>                            if (edit.replacement.equals(SymbolList.EMPTY_LIST) &amp;&amp; exon.contains(edit.pos)) {<a name="line.186"></a>
<span class="sourceLineNo">187</span>                                exon.sequence.edit(new Edit(edit.pos + offset, 0, edit.replacement));<a name="line.187"></a>
<span class="sourceLineNo">188</span>                                offset += edit.length;<a name="line.188"></a>
<span class="sourceLineNo">189</span>                            }<a name="line.189"></a>
<span class="sourceLineNo">190</span>                        }<a name="line.190"></a>
<span class="sourceLineNo">191</span>                    }<a name="line.191"></a>
<span class="sourceLineNo">192</span>                }<a name="line.192"></a>
<span class="sourceLineNo">193</span>            }<a name="line.193"></a>
<span class="sourceLineNo">194</span><a name="line.194"></a>
<span class="sourceLineNo">195</span>            // todo: improve this data structure<a name="line.195"></a>
<span class="sourceLineNo">196</span>            Map&lt;String, ListMultimap&lt;Integer, Allele&gt;&gt; contigs = new HashMap&lt;String, ListMultimap&lt;Integer, Allele&gt;&gt;();<a name="line.196"></a>
<span class="sourceLineNo">197</span><a name="line.197"></a>
<span class="sourceLineNo">198</span>            for (int index : regions.keySet()) {<a name="line.198"></a>
<span class="sourceLineNo">199</span>                for (Allele allele : regions.get(index)) {<a name="line.199"></a>
<span class="sourceLineNo">200</span>                    int sequenceLength = allele.sequence.seqString().length();<a name="line.200"></a>
<span class="sourceLineNo">201</span>                    List&lt;String&gt; fields = Splitter.on("|").splitToList(allele.getName());<a name="line.201"></a>
<span class="sourceLineNo">202</span><a name="line.202"></a>
<span class="sourceLineNo">203</span>                    Double locusLength = Double.parseDouble(fields.get(fields.size() - 1));<a name="line.203"></a>
<span class="sourceLineNo">204</span><a name="line.204"></a>
<span class="sourceLineNo">205</span>                    if (sequenceLength/locusLength &gt;= minimumBreadth) {<a name="line.205"></a>
<span class="sourceLineNo">206</span>                        if (!contigs.containsKey(fields.get(0))) {<a name="line.206"></a>
<span class="sourceLineNo">207</span>                            contigs.put(fields.get(0), ArrayListMultimap.&lt;Integer, Allele&gt;create());<a name="line.207"></a>
<span class="sourceLineNo">208</span>                        }<a name="line.208"></a>
<span class="sourceLineNo">209</span>                        contigs.get(fields.get(0)).put(index, allele);<a name="line.209"></a>
<span class="sourceLineNo">210</span><a name="line.210"></a>
<span class="sourceLineNo">211</span>                        if (!cdna) {<a name="line.211"></a>
<span class="sourceLineNo">212</span>                            writer.println(allele.getName() + "|" + sequenceLength);<a name="line.212"></a>
<span class="sourceLineNo">213</span>                            writer.println(allele.sequence.seqString().toUpperCase());<a name="line.213"></a>
<span class="sourceLineNo">214</span>                        }<a name="line.214"></a>
<span class="sourceLineNo">215</span>                    }<a name="line.215"></a>
<span class="sourceLineNo">216</span>                }<a name="line.216"></a>
<span class="sourceLineNo">217</span>            }<a name="line.217"></a>
<span class="sourceLineNo">218</span><a name="line.218"></a>
<span class="sourceLineNo">219</span>            if (cdna) {<a name="line.219"></a>
<span class="sourceLineNo">220</span>                Map&lt;String, String&gt; cdnas = new HashMap&lt;String, String&gt;();<a name="line.220"></a>
<span class="sourceLineNo">221</span>                for (String contig : contigs.keySet()) {<a name="line.221"></a>
<span class="sourceLineNo">222</span>                    StringBuilder sb = new StringBuilder();<a name="line.222"></a>
<span class="sourceLineNo">223</span><a name="line.223"></a>
<span class="sourceLineNo">224</span>                    for (int index : contigs.get(contig).keySet()) {<a name="line.224"></a>
<span class="sourceLineNo">225</span>                        List&lt;Allele&gt; list = contigs.get(contig).get(index);<a name="line.225"></a>
<span class="sourceLineNo">226</span>                        Collections.sort(list, new Comparator&lt;Allele&gt;() {<a name="line.226"></a>
<span class="sourceLineNo">227</span>                                @Override<a name="line.227"></a>
<span class="sourceLineNo">228</span>                                public int compare(final Allele first, final Allele second) {<a name="line.228"></a>
<span class="sourceLineNo">229</span>                                    return first.sequence.seqString().length() - second.sequence.seqString().length();<a name="line.229"></a>
<span class="sourceLineNo">230</span>                                }<a name="line.230"></a>
<span class="sourceLineNo">231</span>                            });<a name="line.231"></a>
<span class="sourceLineNo">232</span>                        Allele best = list.get(0);<a name="line.232"></a>
<span class="sourceLineNo">233</span>                        sb.append(best.sequence.seqString());<a name="line.233"></a>
<span class="sourceLineNo">234</span>                    }<a name="line.234"></a>
<span class="sourceLineNo">235</span><a name="line.235"></a>
<span class="sourceLineNo">236</span>                    String cdnaSequence = sb.toString();<a name="line.236"></a>
<span class="sourceLineNo">237</span>                    if (removeGaps) {<a name="line.237"></a>
<span class="sourceLineNo">238</span>                        cdnaSequence = cdnaSequence.replaceAll("-", "");<a name="line.238"></a>
<span class="sourceLineNo">239</span>                    }<a name="line.239"></a>
<span class="sourceLineNo">240</span><a name="line.240"></a>
<span class="sourceLineNo">241</span>                    // todo:  use strand from genomic region file<a name="line.241"></a>
<span class="sourceLineNo">242</span>                    if (!(gene.equals("HLA-A") || gene.equals("HLA-DPB1"))) {<a name="line.242"></a>
<span class="sourceLineNo">243</span>                        cdnaSequence = DNATools.reverseComplement(DNATools.createDNA(cdnaSequence)).seqString();<a name="line.243"></a>
<span class="sourceLineNo">244</span>                    }<a name="line.244"></a>
<span class="sourceLineNo">245</span><a name="line.245"></a>
<span class="sourceLineNo">246</span>                    cdnas.put(contig, cdnaSequence.toUpperCase());<a name="line.246"></a>
<span class="sourceLineNo">247</span>                }<a name="line.247"></a>
<span class="sourceLineNo">248</span>                <a name="line.248"></a>
<span class="sourceLineNo">249</span>                List&lt;Map.Entry&lt;String, String&gt;&gt; list = new ArrayList&lt;Map.Entry&lt;String, String&gt;&gt;(cdnas.entrySet());<a name="line.249"></a>
<span class="sourceLineNo">250</span>                Collections.sort(list, new Comparator&lt;Map.Entry&lt;String, String&gt;&gt;() {<a name="line.250"></a>
<span class="sourceLineNo">251</span>                    @Override<a name="line.251"></a>
<span class="sourceLineNo">252</span>                    public int compare(final Map.Entry&lt;String, String&gt; first, final Map.Entry&lt;String, String&gt; second) {<a name="line.252"></a>
<span class="sourceLineNo">253</span>                        return second.getValue().length() - first.getValue().length();<a name="line.253"></a>
<span class="sourceLineNo">254</span>                    }<a name="line.254"></a>
<span class="sourceLineNo">255</span>                });<a name="line.255"></a>
<span class="sourceLineNo">256</span>                <a name="line.256"></a>
<span class="sourceLineNo">257</span>                for (Map.Entry&lt;String, String&gt; entry : list.subList(0, expectedPloidy &gt; list.size() ? list.size() : expectedPloidy)) {<a name="line.257"></a>
<span class="sourceLineNo">258</span>                  writer.println(entry.getKey() + "\n" + entry.getValue());<a name="line.258"></a>
<span class="sourceLineNo">259</span>                }<a name="line.259"></a>
<span class="sourceLineNo">260</span>            }<a name="line.260"></a>
<span class="sourceLineNo">261</span><a name="line.261"></a>
<span class="sourceLineNo">262</span>            return 0;<a name="line.262"></a>
<span class="sourceLineNo">263</span>        }<a name="line.263"></a>
<span class="sourceLineNo">264</span>        finally {<a name="line.264"></a>
<span class="sourceLineNo">265</span>            try {<a name="line.265"></a>
<span class="sourceLineNo">266</span>                writer.close();<a name="line.266"></a>
<span class="sourceLineNo">267</span>            }<a name="line.267"></a>
<span class="sourceLineNo">268</span>            catch (Exception e) {<a name="line.268"></a>
<span class="sourceLineNo">269</span>                // empty<a name="line.269"></a>
<span class="sourceLineNo">270</span>            }<a name="line.270"></a>
<span class="sourceLineNo">271</span>        }<a name="line.271"></a>
<span class="sourceLineNo">272</span>    }<a name="line.272"></a>
<span class="sourceLineNo">273</span><a name="line.273"></a>
<span class="sourceLineNo">274</span><a name="line.274"></a>
<span class="sourceLineNo">275</span>    static List&lt;Edit&gt; cigarToEditList(final SAMRecord record) throws IllegalSymbolException {<a name="line.275"></a>
<span class="sourceLineNo">276</span>        int position = 0;<a name="line.276"></a>
<span class="sourceLineNo">277</span>        List&lt;Edit&gt; edits = new ArrayList&lt;Edit&gt;();<a name="line.277"></a>
<span class="sourceLineNo">278</span>        for (CigarElement element : record.getCigar().getCigarElements()) {<a name="line.278"></a>
<span class="sourceLineNo">279</span>            position += element.getLength();<a name="line.279"></a>
<span class="sourceLineNo">280</span><a name="line.280"></a>
<span class="sourceLineNo">281</span>            String operator = element.getOperator().toString();<a name="line.281"></a>
<span class="sourceLineNo">282</span>            if ("I".equals(operator) || "S".equals(operator)) {<a name="line.282"></a>
<span class="sourceLineNo">283</span>                edits.add(new Edit(position - element.getLength() + 1, element.getLength(), SymbolList.EMPTY_LIST));<a name="line.283"></a>
<span class="sourceLineNo">284</span>                position -= element.getLength();<a name="line.284"></a>
<span class="sourceLineNo">285</span>            }<a name="line.285"></a>
<span class="sourceLineNo">286</span>            else if ("D".equals(operator)) {<a name="line.286"></a>
<span class="sourceLineNo">287</span>                SymbolList replace = DNATools.createDNA(Joiner.on("").join(Collections.nCopies(element.getLength(), "-")));<a name="line.287"></a>
<span class="sourceLineNo">288</span>                edits.add(new Edit(position + element.getLength() - 1, 0, replace));<a name="line.288"></a>
<span class="sourceLineNo">289</span>            }<a name="line.289"></a>
<span class="sourceLineNo">290</span>        }<a name="line.290"></a>
<span class="sourceLineNo">291</span>        return edits;<a name="line.291"></a>
<span class="sourceLineNo">292</span>    }<a name="line.292"></a>
<span class="sourceLineNo">293</span><a name="line.293"></a>
<span class="sourceLineNo">294</span>    // todo:  use BED format; extract BED format reader from e.g. LiftoverBed<a name="line.294"></a>
<span class="sourceLineNo">295</span>    static Map&lt;Integer, Allele&gt; readGenomicFile(final File inputGenomicFile) throws IOException {<a name="line.295"></a>
<span class="sourceLineNo">296</span>        int lineNumber = 0;<a name="line.296"></a>
<span class="sourceLineNo">297</span>        BufferedReader reader = null;<a name="line.297"></a>
<span class="sourceLineNo">298</span>        Map&lt;Integer, Allele&gt; exons = new HashMap&lt;Integer, Allele&gt;();<a name="line.298"></a>
<span class="sourceLineNo">299</span>        try {<a name="line.299"></a>
<span class="sourceLineNo">300</span>            reader = reader(inputGenomicFile);<a name="line.300"></a>
<span class="sourceLineNo">301</span><a name="line.301"></a>
<span class="sourceLineNo">302</span>            while (reader.ready()) {<a name="line.302"></a>
<span class="sourceLineNo">303</span>                String line = reader.readLine();<a name="line.303"></a>
<span class="sourceLineNo">304</span>                if (line == null) {<a name="line.304"></a>
<span class="sourceLineNo">305</span>                    break;<a name="line.305"></a>
<span class="sourceLineNo">306</span>                }<a name="line.306"></a>
<span class="sourceLineNo">307</span>                String[] tokens = line.split("\\s+");<a name="line.307"></a>
<span class="sourceLineNo">308</span>                if (tokens.length != 2) {<a name="line.308"></a>
<span class="sourceLineNo">309</span>                    throw new IOException("invalid genomic file format at line " + lineNumber + ", invalid number of tokens");<a name="line.309"></a>
<span class="sourceLineNo">310</span>                }<a name="line.310"></a>
<span class="sourceLineNo">311</span><a name="line.311"></a>
<span class="sourceLineNo">312</span>                try {<a name="line.312"></a>
<span class="sourceLineNo">313</span>                    int exon = Integer.parseInt(tokens[0]);<a name="line.313"></a>
<span class="sourceLineNo">314</span>                    Locus locus = FeatureParser.parseLocus(tokens[1]);<a name="line.314"></a>
<span class="sourceLineNo">315</span>                    exons.put(exon, Allele.builder()<a name="line.315"></a>
<span class="sourceLineNo">316</span>                              .withContig(locus.getContig())<a name="line.316"></a>
<span class="sourceLineNo">317</span>                              .withStart(locus.getMin())<a name="line.317"></a>
<span class="sourceLineNo">318</span>                              .withEnd(locus.getMax() + 1)<a name="line.318"></a>
<span class="sourceLineNo">319</span>                              .build());<a name="line.319"></a>
<span class="sourceLineNo">320</span>                }<a name="line.320"></a>
<span class="sourceLineNo">321</span>                catch (NumberFormatException | AlleleException | ParseException e) {<a name="line.321"></a>
<span class="sourceLineNo">322</span>                    throw new IOException("invalid genomic file format at line " + lineNumber + ", caught " + e.getMessage());<a name="line.322"></a>
<span class="sourceLineNo">323</span>                }<a name="line.323"></a>
<span class="sourceLineNo">324</span><a name="line.324"></a>
<span class="sourceLineNo">325</span>                lineNumber++;<a name="line.325"></a>
<span class="sourceLineNo">326</span>            }<a name="line.326"></a>
<span class="sourceLineNo">327</span>        }<a name="line.327"></a>
<span class="sourceLineNo">328</span>        finally {<a name="line.328"></a>
<span class="sourceLineNo">329</span>            try {<a name="line.329"></a>
<span class="sourceLineNo">330</span>                reader.close();<a name="line.330"></a>
<span class="sourceLineNo">331</span>            }<a name="line.331"></a>
<span class="sourceLineNo">332</span>            catch (Exception e) {<a name="line.332"></a>
<span class="sourceLineNo">333</span>                // ignore<a name="line.333"></a>
<span class="sourceLineNo">334</span>            }<a name="line.334"></a>
<span class="sourceLineNo">335</span>        }<a name="line.335"></a>
<span class="sourceLineNo">336</span>        return exons;<a name="line.336"></a>
<span class="sourceLineNo">337</span>    }<a name="line.337"></a>
<span class="sourceLineNo">338</span><a name="line.338"></a>
<span class="sourceLineNo">339</span><a name="line.339"></a>
<span class="sourceLineNo">340</span>    /**<a name="line.340"></a>
<span class="sourceLineNo">341</span>     * Main.<a name="line.341"></a>
<span class="sourceLineNo">342</span>     *<a name="line.342"></a>
<span class="sourceLineNo">343</span>     * @param args command line args<a name="line.343"></a>
<span class="sourceLineNo">344</span>     */<a name="line.344"></a>
<span class="sourceLineNo">345</span>    public static void main(final String[] args) {<a name="line.345"></a>
<span class="sourceLineNo">346</span>        Switch about = new Switch("a", "about", "display about message");<a name="line.346"></a>
<span class="sourceLineNo">347</span>        Switch help = new Switch("h", "help", "display help message");<a name="line.347"></a>
<span class="sourceLineNo">348</span>        FileArgument inputBamFile = new FileArgument("i", "input-bam-file", "input BAM file", true);<a name="line.348"></a>
<span class="sourceLineNo">349</span>        FileArgument inputGenomicFile = new FileArgument("x", "input-genomic-range-file", "input file of genomic ranges, space-delimited exon chrom:start-end", true);<a name="line.349"></a>
<span class="sourceLineNo">350</span>        FileArgument outputFile = new FileArgument("o", "output-file", "output FASTA file, default stdout", false);<a name="line.350"></a>
<span class="sourceLineNo">351</span>        StringArgument gene = new StringArgument("g", "gene", "gene name, written to the FASTA headers", true);<a name="line.351"></a>
<span class="sourceLineNo">352</span>        Switch cdna = new Switch("c", "cdna", "output cDNA from the same contig (phased consensus sequence) in FASTA format");<a name="line.352"></a>
<span class="sourceLineNo">353</span>        Switch removeGaps = new Switch("r", "remove-gaps", "remove alignment gaps in the filtered consensus sequence");<a name="line.353"></a>
<span class="sourceLineNo">354</span>        DoubleArgument minimumBreadth = new DoubleArgument("b", "minimum-breadth-of-coverage", "filter contigs less than minimum, default " + DEFAULT_MINIMUM_BREADTH, false);<a name="line.354"></a>
<span class="sourceLineNo">355</span>        IntegerArgument expectedPloidy = new IntegerArgument("p", "expected-ploidy", "filter contigs more than expected ploidy, default " + DEFAULT_EXPECTED_PLOIDY, false);<a name="line.355"></a>
<span class="sourceLineNo">356</span><a name="line.356"></a>
<span class="sourceLineNo">357</span>        ArgumentList arguments = new ArgumentList(about, help, inputBamFile, inputGenomicFile, outputFile, gene, cdna, removeGaps, minimumBreadth, expectedPloidy);<a name="line.357"></a>
<span class="sourceLineNo">358</span>        CommandLine commandLine = new CommandLine(args);<a name="line.358"></a>
<span class="sourceLineNo">359</span><a name="line.359"></a>
<span class="sourceLineNo">360</span>        FilterConsensus filterConsensus = null;<a name="line.360"></a>
<span class="sourceLineNo">361</span>        try<a name="line.361"></a>
<span class="sourceLineNo">362</span>        {<a name="line.362"></a>
<span class="sourceLineNo">363</span>            CommandLineParser.parse(commandLine, arguments);<a name="line.363"></a>
<span class="sourceLineNo">364</span>            if (about.wasFound()) {<a name="line.364"></a>
<span class="sourceLineNo">365</span>                About.about(System.out);<a name="line.365"></a>
<span class="sourceLineNo">366</span>                System.exit(0);<a name="line.366"></a>
<span class="sourceLineNo">367</span>            }<a name="line.367"></a>
<span class="sourceLineNo">368</span>            if (help.wasFound()) {<a name="line.368"></a>
<span class="sourceLineNo">369</span>                Usage.usage(USAGE, null, commandLine, arguments, System.out);<a name="line.369"></a>
<span class="sourceLineNo">370</span>                System.exit(0);<a name="line.370"></a>
<span class="sourceLineNo">371</span>            }<a name="line.371"></a>
<span class="sourceLineNo">372</span><a name="line.372"></a>
<span class="sourceLineNo">373</span>            if (!inputBamFile.getValue().exists()) {<a name="line.373"></a>
<span class="sourceLineNo">374</span>                throw new IllegalArgumentException("-i, --input-bam-file must be a file that exists");<a name="line.374"></a>
<span class="sourceLineNo">375</span>            }<a name="line.375"></a>
<span class="sourceLineNo">376</span>            if (!inputGenomicFile.getValue().exists()) {<a name="line.376"></a>
<span class="sourceLineNo">377</span>                throw new IllegalArgumentException("-x, --input-genomic-range-file must be a file that exists");<a name="line.377"></a>
<span class="sourceLineNo">378</span>            }<a name="line.378"></a>
<span class="sourceLineNo">379</span><a name="line.379"></a>
<span class="sourceLineNo">380</span>            filterConsensus = new FilterConsensus(inputBamFile.getValue(),<a name="line.380"></a>
<span class="sourceLineNo">381</span>                                                  inputGenomicFile.getValue(),<a name="line.381"></a>
<span class="sourceLineNo">382</span>                                                  outputFile.getValue(),<a name="line.382"></a>
<span class="sourceLineNo">383</span>                                                  gene.getValue(),<a name="line.383"></a>
<span class="sourceLineNo">384</span>                                                  cdna.wasFound(),<a name="line.384"></a>
<span class="sourceLineNo">385</span>                                                  removeGaps.wasFound(),<a name="line.385"></a>
<span class="sourceLineNo">386</span>                                                  minimumBreadth.getValue(DEFAULT_MINIMUM_BREADTH),<a name="line.386"></a>
<span class="sourceLineNo">387</span>                                                  expectedPloidy.getValue(DEFAULT_EXPECTED_PLOIDY));<a name="line.387"></a>
<span class="sourceLineNo">388</span>        }<a name="line.388"></a>
<span class="sourceLineNo">389</span>        catch (CommandLineParseException | IllegalArgumentException e) {<a name="line.389"></a>
<span class="sourceLineNo">390</span>            if (about.wasFound()) {<a name="line.390"></a>
<span class="sourceLineNo">391</span>                About.about(System.out);<a name="line.391"></a>
<span class="sourceLineNo">392</span>                System.exit(0);<a name="line.392"></a>
<span class="sourceLineNo">393</span>            }<a name="line.393"></a>
<span class="sourceLineNo">394</span>            if (help.wasFound()) {<a name="line.394"></a>
<span class="sourceLineNo">395</span>                Usage.usage(USAGE, null, commandLine, arguments, System.out);<a name="line.395"></a>
<span class="sourceLineNo">396</span>                System.exit(0);<a name="line.396"></a>
<span class="sourceLineNo">397</span>            }<a name="line.397"></a>
<span class="sourceLineNo">398</span>            Usage.usage(USAGE, e, commandLine, arguments, System.err);<a name="line.398"></a>
<span class="sourceLineNo">399</span>            System.exit(-1);<a name="line.399"></a>
<span class="sourceLineNo">400</span>        }<a name="line.400"></a>
<span class="sourceLineNo">401</span>        try {<a name="line.401"></a>
<span class="sourceLineNo">402</span>            System.exit(filterConsensus.call());<a name="line.402"></a>
<span class="sourceLineNo">403</span>        }<a name="line.403"></a>
<span class="sourceLineNo">404</span>        catch (Exception e) {<a name="line.404"></a>
<span class="sourceLineNo">405</span>            e.printStackTrace();<a name="line.405"></a>
<span class="sourceLineNo">406</span>            System.exit(1);<a name="line.406"></a>
<span class="sourceLineNo">407</span>        }<a name="line.407"></a>
<span class="sourceLineNo">408</span>    }<a name="line.408"></a>
<span class="sourceLineNo">409</span>}<a name="line.409"></a>




























































</pre>
</div>
</body>
</html>
